sql,kv: handle blind UPSERTs and DELETEs with buffered writes

yuzefovich opened the pull request:
**sql: disable buffered writes for COPY**

COPY machinery performs bulk writes without doing any reads, and it already performs batching itself, so there is no benefit from enabling buffered writes for it.

**sql,kv: add MustAcquireExclusiveLock flag to Put and use it for blind UPSERT**

This commit extends `PutRequest` with `MustAcquireExclusiveLock` flag that indicates that a lock with the strength no lower than "exclusive" should be acquired on the key, even if the key doesn't exist yet. This ability is then utilized by the SQL layer when performing blind UPSERTs (which is the case where we can avoid the initial scan, in some special cases). (We do that unconditionally, regardless of buffered writes being enabled, since the write intent is a stronger lock.)

Additionally, this commit does a minor cleanup of `InsertRow` interface to take an explicit value which KV write operation should be performed.

**kv: remove some duplication with DelRange**

**kv: extend Del request for buffered writes**

This commit adds `MustAcquireExclusiveLock` flag to the `DeleteRequest` that the SQL layer will use to communicate to the KV client that locks need to be acquired on the keys when writes are buffered on the KV client. This flag is now utilized in the "blind" fast-path DELETE case where we skip the initial scan. All other usages of Del will be audited separately.

There are two alternatives to this approach that I like less:
- we could reuse existing `ReturnKeys` flag of the `DeleteRange` request (and add a similar flag to the `Delete` request) to mean that locks are needed. I don't like it as it conflates different purposes.
- we could adjust the SQL layer to explicitly issue Gets and Scans. I don't like this since SQL doesn't really need the values stored at the keys being deleted.

For now DeleteRange requests in the blind DELETE fast-path are left as a TODO since we haven't fully committed to buffering them in the interceptor.

Fixes: #139105.
Epic: CRDB-45273
Release note: None

michae2 reviewed (commented):
> Reviewed 1 of 19 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140158)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @stevendanna, and @yuzefovich)

___
*[`pkg/sql/tablewriter_upsert.go` line 179 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIIVpRB4vvDv2yrij1N:-OIIVpRB4vvDv2yrij1O:b-cosdfw) ([raw file](https://github.com/cockroachdb/cockroach/blob/70409972e272f70cd71ee2eb47ba93b4cd5510a4/pkg/sql/tablewriter_upsert.go#L179)):*
> ```Go
> 			// TODO(yuzefovich): add a tracing test for this once buffered
> 			// writes are more fleshed out.
> 			overwrite = false
> ```

Hmm, now that I think about it: wouldn't directly replacing Put with CPut cause blind UPSERT to fail when the row already exists? The beauty of Put is that it doesn't read the existing value... CPut does...

:thinking:

Maybe we need either (a) a version of CPut / CPutAllowingIfNotExists that doesn't actually read the existing value or (b) a flag on Put that says "also acquire a lock"?


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140158)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @stevendanna, and @yuzefovich)

___
*[`pkg/sql/tablewriter_upsert.go` line 179 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIIVpRB4vvDv2yrij1N:-OIIZ6xr8bXw-ThW08tq:b-p79pyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/70409972e272f70cd71ee2eb47ba93b4cd5510a4/pkg/sql/tablewriter_upsert.go#L179)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Hmm, now that I think about it: wouldn't directly replacing Put with CPut cause blind UPSERT to fail when the row already exists? The beauty of Put is that it doesn't read the existing value... CPut does...

:thinking:

Maybe we need either (a) a version of CPut / CPutAllowingIfNotExists that doesn't actually read the existing value or (b) a flag on Put that says "also acquire a lock"?
</blockquote></details>

(b) seems kinda like what you're proposing for Del and DelRange in https://github.com/cockroachdb/cockroach/pull/140185/commits/e7d1a677736bceb3194a0242563fbbb64f496784


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140158)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @michae2, and @stevendanna)

___
*[`pkg/sql/tablewriter_upsert.go` line 179 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIIVpRB4vvDv2yrij1N:-OIIwSyA1e9EQy-CoPz8:b-tyjuz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/70409972e272f70cd71ee2eb47ba93b4cd5510a4/pkg/sql/tablewriter_upsert.go#L179)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

(b) seems kinda like what you're proposing for Del and DelRange in https://github.com/cockroachdb/cockroach/pull/140185/commits/e7d1a677736bceb3194a0242563fbbb64f496784
</blockquote></details>

Yeah, that's a good point that I missed. Indeed, (b) sounds similar to what I have in flight for Del and DelRange. Perhaps let me polish that change up for review, and if that approach is accepted, we'll use it - "it" meaning the new flag on Put - in this PR too. I like (b) over (a) because _by definition_ CPut is a conditional Put which _should_ read the existing value, so adding a flag to CPut to not read it seems confusing.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140158)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @michae2, and @yuzefovich)

___
*[`pkg/sql/tablewriter_upsert.go` line 179 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIIVpRB4vvDv2yrij1N:-OIKXTin789N9uzs6DCO:b-ln0kkb) ([raw file](https://github.com/cockroachdb/cockroach/blob/70409972e272f70cd71ee2eb47ba93b4cd5510a4/pkg/sql/tablewriter_upsert.go#L179)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Yeah, that's a good point that I missed. Indeed, (b) sounds similar to what I have in flight for Del and DelRange. Perhaps let me polish that change up for review, and if that approach is accepted, we'll use it - "it" meaning the new flag on Put - in this PR too. I like (b) over (a) because _by definition_ CPut is a conditional Put which _should_ read the existing value, so adding a flag to CPut to not read it seems confusing.
</blockquote></details>

I think overloading "CPut" has a long tradition in software engineer. But, I agree with y'all that we should avoid doing it here if we can.

One thing we should keep an eye on is that we don't end up making the implicit transaction case work for these fast paths. In the txn interceptor, we should make sure we recognise the case  where we can elide the lock acquisition.  


<!-- Sent from Reviewable.io -->

yuzefovich committed:
sql: disable buffered writes for COPY

COPY machinery performs bulk writes without doing any reads, and it
already performs batching itself, so there is no benefit from enabling
buffered writes for it.

Release note: None

yuzefovich committed:
sql,kv: add MustAcquireExclusiveLock flag to Put and use it for blind UPSERT

This commit extends `PutRequest` with `MustAcquireExclusiveLock` flag that
indicates that a lock with the strength no lower than "exclusive" should
be acquired on the key, even if the key doesn't exist yet. This ability
is then utilized by the SQL layer when performing blind UPSERTs (which
is the case where we can avoid the initial scan, in some special cases).
(We do that unconditionally, regardless of buffered writes being
enabled, since the write intent is a stronger lock.)

Additionally, this commit does a minor cleanup of `InsertRow` interface to
take an explicit value which KV write operation should be performed.

Release note: None

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140158)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @kev-cao, @michae2, and @stevendanna)

___
*[`pkg/sql/tablewriter_upsert.go` line 179 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIIVpRB4vvDv2yrij1N:-OIXM6he8XtEO0H5UiGT:b-bcbnms) ([raw file](https://github.com/cockroachdb/cockroach/blob/70409972e272f70cd71ee2eb47ba93b4cd5510a4/pkg/sql/tablewriter_upsert.go#L179)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

I think overloading "CPut" has a long tradition in software engineer. But, I agree with y'all that we should avoid doing it here if we can.

One thing we should keep an eye on is that we don't end up making the implicit transaction case work for these fast paths. In the txn interceptor, we should make sure we recognise the case  where we can elide the lock acquisition.
</blockquote></details>

Modified the PR to add a boolean flag to Put. Also added a TODO for the Steven's callout. PTAL.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> Reviewed 33 of 33 files at r2, 15 of 15 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140158)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao, @michae2, @stevendanna, and @yuzefovich)

___
*[`pkg/sql/tablewriter_upsert.go` line 174 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIXokmh-9LhrHuVvSxH:-OIXokmh-9LhrHuVvSxI:bcmtiti) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9abcdc351df6d48983297429e9de408ac8bf562/pkg/sql/tablewriter_upsert.go#L174)):*
> ```Go
> 		// we're in the UPSERT fast-path).
> 		kvOp := row.PutOp
> 		if tu.bufferedWritesEnabled {
> ```

Could we set `row.PutLockingOp` unconditionally? Without buffered writes, we're writing an intent, and we consider `Intent` locking strength to be stronger than `Exclusive` -- so, from KVs perspective, we are respecting the contract defined on the flag you added.
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 275 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIXlX2XDKF6mBXi4lne:-OIXlX2XDKF6mBXi4lnf:ban7jmn) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9abcdc351df6d48983297429e9de408ac8bf562/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L275)):*
> ```Go
> 			// we need to add a locking Get to the BatchRequest, including if
> 			// the key doesn't exist (#139232).
> 			// TODO(yuzefovich): ensure that we elide the lock acquisition
> ```

The `txnPipeliner` already keeps track of all the locks that have been acquired by a transaction. It would be nice to consult it here to elide certain lock acquisitions -- I have a TODO in our tracking spreadsheet to file an issue to this effect. 

It might be hard to fit this all together code design wise, though, because the interceptors here are mostly agnostic to other interceptors in the stack[1].

Note that we compress the `lockFootprint` struct on the pipeliner, so we won't be able to consult it in those cases.

---- 

[1] One idea I had was that we could elide the request in the pipeliner, instead of the write buffer, if the pipeliner notices this flag and it has already acquired a lock on the key. The that only works if we know that the write was buffered by the `txnWriteBuffer`, which we don't.
___
*[`pkg/kv/kvpb/api.proto` line 290 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIXnlgy2tCQ11jUF6Wd:-OIXnlgy2tCQ11jUF6We:b-xwvz77) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9abcdc351df6d48983297429e9de408ac8bf562/pkg/kv/kvpb/api.proto#L290)):*
> ```proto
>   // than "exclusive" needs to be acquired on the key, even if it doesn't exist.
>   // At the moment, this is fully handled by the KV client.
>   bool lock_exclusive = 5;
> ```

How do you feel about naming this `MustAcquireExclusiveLock` instead to make the contract clearer? 


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140158)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @kev-cao, @michae2, and @stevendanna)

___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 275 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIXlX2XDKF6mBXi4lne:-OIgjQ27C7I5lg3zDN_c:bfz20uf) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9abcdc351df6d48983297429e9de408ac8bf562/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L275)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

The `txnPipeliner` already keeps track of all the locks that have been acquired by a transaction. It would be nice to consult it here to elide certain lock acquisitions -- I have a TODO in our tracking spreadsheet to file an issue to this effect. 

It might be hard to fit this all together code design wise, though, because the interceptors here are mostly agnostic to other interceptors in the stack[1].

Note that we compress the `lockFootprint` struct on the pipeliner, so we won't be able to consult it in those cases.

---- 

[1] One idea I had was that we could elide the request in the pipeliner, instead of the write buffer, if the pipeliner notices this flag and it has already acquired a lock on the key. The that only works if we know that the write was buffered by the `txnWriteBuffer`, which we don't.
</blockquote></details>

Thanks for the context.

This TODO is about a slightly different case though that Steven highlighted: namely, with the blind UPSERT issued in an implicit txn if we add a locking Get stage with buffered writes enabled, then we would regress in comparison to the non-buffered writes case.
___
*[`pkg/kv/kvpb/api.proto` line 290 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIXnlgy2tCQ11jUF6Wd:-OIgfNe03nv0pRT25fyU:b-86mr6o) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9abcdc351df6d48983297429e9de408ac8bf562/pkg/kv/kvpb/api.proto#L290)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

How do you feel about naming this `MustAcquireExclusiveLock` instead to make the contract clearer?
</blockquote></details>

I like it, done.
___
*[`pkg/sql/tablewriter_upsert.go` line 174 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIXokmh-9LhrHuVvSxH:-OIggC413UBv-sR0NSKs:be1hhbj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9abcdc351df6d48983297429e9de408ac8bf562/pkg/sql/tablewriter_upsert.go#L174)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Could we set `row.PutLockingOp` unconditionally? Without buffered writes, we're writing an intent, and we consider `Intent` locking strength to be stronger than `Exclusive` -- so, from KVs perspective, we are respecting the contract defined on the flag you added.
</blockquote></details>

Yes, we could. I wasn't sure whether it'd be cleaner or not that way, but I don't see a reason to not do it, done.


<!-- Sent from Reviewable.io -->

stevendanna reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140158)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @michae2)

___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 275 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIXlX2XDKF6mBXi4lne:-OIjfinB4QjYCjI_aDNQ:b-abya3s) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9abcdc351df6d48983297429e9de408ac8bf562/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L275)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Thanks for the context.

This TODO is about a slightly different case though that Steven highlighted: namely, with the blind UPSERT issued in an implicit txn if we add a locking Get stage with buffered writes enabled, then we would regress in comparison to the non-buffered writes case.
</blockquote></details>

Now that we see this coming together, perhaps the implicit transaction case is handled transparently by the fact that if the EndTxn is in the same batch then we don't need to do any transformations of other requests in the batch.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140158)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @michae2, and @stevendanna)

___
*[`pkg/sql/tablewriter_upsert.go` line 174 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIXokmh-9LhrHuVvSxH:-OIlovSqDl3ZiC08jAWz:b-o77ba) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9abcdc351df6d48983297429e9de408ac8bf562/pkg/sql/tablewriter_upsert.go#L174)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Yes, we could. I wasn't sure whether it'd be cleaner or not that way, but I don't see a reason to not do it, done.
</blockquote></details>

The CI failure shows a reason for why we might consider using `row.PutOp` when buffered writes are disabled - the tracing message would be different: if we use the locking Put unconditionally, then we might see `Put (locking) ...`, which might be confusing. Not sure which side I lean towards.


<!-- Sent from Reviewable.io -->

yuzefovich committed:
kv: remove some duplication with DelRange

Release note: None

yuzefovich committed:
kv: extend Del request for buffered writes

This commit adds `MustAcquireExclusiveLock` flag to the `DeleteRequest`
that the SQL layer will use to communicate to the KV client that locks
need to be acquired on the keys when writes are buffered on the KV
client. This flag is now utilized in the "blind" fast-path DELETE case
where we skip the initial scan. All other usages of Del will be audited
separately.

There are two alternatives to this approach that I like less:
- we could reuse existing `ReturnKeys` flag of the `DeleteRange` request
(and add a similar flag to the `Delete` request) to mean that locks are
needed. I don't like it as it conflates different purposes.
- we could adjust the SQL layer to explicitly issue Gets and Scans.
I don't like this since SQL doesn't really need the values stored at the
keys being deleted.

For now DeleteRange requests in the blind DELETE fast-path are left as
a TODO since we haven't fully committed to buffering them in the
interceptor.

Release note: None

yuzefovich reviewed (commented):
I decided to include the blind DELETE fast-path handling into this PR since it'll have the same questions (naming of the new boolean on the DeleteRequest; whether we use it unconditionally in the blind fast-path code path).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-:-OImu7vQ57U3Or2z1T-Q:by35vuh)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @michae2, and @stevendanna)



<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Very nice!

I think we're also going to need to set these flags for Puts and Dels issued by UPDATE statements that don't use an initial locking scan, but that can be in another PR.

> Reviewed 1 of 33 files at r2, 34 of 34 files at r6, 16 of 16 files at r7, 19 of 19 files at r8, 7 of 7 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-:-OIvYgo9CVW9PtOr2MMX:b-901erc)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/sql/row/updater.go` line 372 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIvTfIPAXNkY_71z9ZJ:-OIvTfIPAXNkY_71z9ZK:blidz36) ([raw file](https://github.com/cockroachdb/cockroach/blob/553ef49ab7bd082504cde3dbd0e07176212d7e3e/pkg/sql/row/updater.go#L372)):*
> ```Go
> 		ru.newValues, ru.FetchColIDtoRowIndex,
> 		ru.UpdateColIDtoRowIndex,
> 		&ru.key, &ru.value, ru.valueBuf, oth, oldValues, PutOp, traceKV)
> ```

Is this the one that will need to be `PutMustAcquireExclusiveLockOp` if the initial scan wasn't locking?
___
*[`pkg/sql/delete_range.go` line 171 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIvYDoZ8OxGdHggjvd7:-OIvYDoZ8OxGdHggjvd8:bilmndj) ([raw file](https://github.com/cockroachdb/cockroach/blob/03823af592b6ffa919a8181e6cc26300ea39912c/pkg/sql/delete_range.go#L171)):*
> ```Go
> 		if span.EndKey == nil {
> 			if traceKV {
> 				log.VEventf(ctx, 2, "Del %s", span.Key)
> ```

might be nice to add "(locking)" to the trace
___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 288 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIvXur2ECUeJOBaomPC:-OIvXur2ECUeJOBaomPD:b-rh4cvp) ([raw file](https://github.com/cockroachdb/cockroach/blob/03823af592b6ffa919a8181e6cc26300ea39912c/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L288)):*
> ```Go
> 			ru.MustSetInner(&kvpb.DeleteResponse{
> 				// TODO(yuzefovich): if MustAcquireExclusiveLock flag is set on
> 				// the Put, then we need to add a locking Get to the
> ```

should this be "Del" instead of "Put"?


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140158)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani and @yuzefovich)

___
*[`pkg/sql/tablewriter_upsert.go` line 174 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIXokmh-9LhrHuVvSxH:-OIvZJy61fyb35SxumXe:bg3iqqa) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9abcdc351df6d48983297429e9de408ac8bf562/pkg/sql/tablewriter_upsert.go#L174)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

The CI failure shows a reason for why we might consider using `row.PutOp` when buffered writes are disabled - the tracing message would be different: if we use the locking Put unconditionally, then we might see `Put (locking) ...`, which might be confusing. Not sure which side I lean towards.
</blockquote></details>

I like the idea of using it unconditionally. I think the (locking) in the trace message is true, we do lock the row.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Yes, good point - I plan to audit them separately (Dels are tracked in #139160).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-:-OIv_8oZ6qBfZXhATwR9:bd0djax)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani, @michae2, and @stevendanna)

___
*[`pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go` line 288 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIvXur2ECUeJOBaomPC:-OIv_aUeAFzc_6AKW5UC:brpnd8q) ([raw file](https://github.com/cockroachdb/cockroach/blob/03823af592b6ffa919a8181e6cc26300ea39912c/pkg/kv/kvclient/kvcoord/txn_interceptor_write_buffer.go#L288)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

should this be "Del" instead of "Put"?
</blockquote></details>

Oops, yes, ty.
___
*[`pkg/sql/delete_range.go` line 171 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIvYDoZ8OxGdHggjvd7:-OIv_ir81k0GlhFJZOzG:bfbq99i) ([raw file](https://github.com/cockroachdb/cockroach/blob/03823af592b6ffa919a8181e6cc26300ea39912c/pkg/sql/delete_range.go#L171)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

might be nice to add "(locking)" to the trace
</blockquote></details>

Good point, done.
___
*[`pkg/sql/tablewriter_upsert.go` line 174 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIXokmh-9LhrHuVvSxH:-OIvb1ow25QIvWv_UmqL:b-pugg5x) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9abcdc351df6d48983297429e9de408ac8bf562/pkg/sql/tablewriter_upsert.go#L174)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

I like the idea of using it unconditionally. I think the (locking) in the trace message is true, we do lock the row.
</blockquote></details>

That's true. My main concern was that - with buffered writes disabled - it now becomes possible to see both "locking" and "non-locking" messages in the trace even though all of them are locking in nature. However, since we do plan to enable buffered writes by default, having a clear distinction of when the SQL requires a lock vs doesn't require it in the trace seems helpful. Trace is mostly for DB engineers after all, so the extra detail - that could be somewhat confusing in some scenarios - seems warranted. Let's keep it then.
___
*[`pkg/sql/row/updater.go` line 372 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIvTfIPAXNkY_71z9ZJ:-OIvaeaP2Hv_T7GEljQ7:b-q1kpqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/553ef49ab7bd082504cde3dbd0e07176212d7e3e/pkg/sql/row/updater.go#L372)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Is this the one that will need to be `PutMustAcquireExclusiveLockOp` if the initial scan wasn't locking?
</blockquote></details>

Yes, I think so.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-:-OIwpuFP2ciMBaqBS5nb:b-tlx7ko)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @arulajmani, @michae2, and @stevendanna)

___
*[`pkg/sql/tablewriter_upsert.go` line 174 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140158#-OIXokmh-9LhrHuVvSxH:-OIwpZqm7HRW9LODnCqe:b-7nucuc) ([raw file](https://github.com/cockroachdb/cockroach/blob/b9abcdc351df6d48983297429e9de408ac8bf562/pkg/sql/tablewriter_upsert.go#L174)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

That's true. My main concern was that - with buffered writes disabled - it now becomes possible to see both "locking" and "non-locking" messages in the trace even though all of them are locking in nature. However, since we do plan to enable buffered writes by default, having a clear distinction of when the SQL requires a lock vs doesn't require it in the trace seems helpful. Trace is mostly for DB engineers after all, so the extra detail - that could be somewhat confusing in some scenarios - seems warranted. Let's keep it then.
</blockquote></details>

Thought I would mention this explicitly - we're ok using these `MustAcquireExclusiveLock` flags unconditionally with regards to mixed version clusters even though we introduced these in 25.2 timeframe since they have no behavior change for pre-25.2 versions where there is no concept of buffered writes.


<!-- Sent from Reviewable.io -->
